### PR TITLE
mkchromecast: init at 0.3.8.1

### PR DESCRIPTION
**Pull request by Shou**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by dtzWill** [comment]
LGTM! Thanks!

* Hopefully they make a new release soon :3
* Do you know how to get notifications working? Adding `pygobject3` and `libnotify` to `propagatedBuildInputs` didn't seem enough.  Not a deal-breaker but thought I'd ask :grin:.

**Comment by Shou** [comment]
I haven't had much success in making notifications work. Someone has a [minimal example](https://github.com/Mic92/themenabend-nixos/blob/master/nix-shebang-python) that mimicks how it's done in `mkchromecast` but if you add `--pure` to that it emits `gi.repository.GLib.Error: g-exec-error-quark: Failed to execute child process “dbus-launch” (No such file or directory) (8)`. The system tray icon still works, it only won't emit notifications during the Chromecast device scan as far as I can tell.

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+, alsaSupport ? false
+, systemTray ? true
+, ffmpegSupport ? true
+, youtubeSupport ? true
+}:
+
+python3Packages.buildPythonApplication rec {
+  pname = "mkchromecast";
+  version = "0.3.8.1";
+  name = "${pname}-${version}";
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
```

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+  src = fetchurl {
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
  src = fetchFromGitHub {
```

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    url = "https://github.com/muammar/${pname}/archive/${version}.tar.gz";
+    sha256 = "1clr02zphpkjnnw5l1mgbjsk5ycqryajnlpb8wfdb5ax72ca7066";
+  };
+
+  buildInputs = [ makeWrapper ];
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
  nativeBuildInputs = [ makeWrapper ];
```

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    flask
+    netifaces
+    requests
+    pkgs.vorbis-tools
+    pkgs.sox
+    pkgs.flac
+    pkgs.faac
+    pkgs.lame
+    pkgs.ffmpeg
+    pkgs.nodejs
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]: Dependencies that aren't Python packages should be in `buildInputs` and they should be specified as arguments. However, you can probably leave them out anyway because you need to patch commands such as https://github.com/muammar/mkchromecast/blob/0.3.8.1/mkchromecast/video.py#L61 using `substituteAll`.

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 5]
```diff
@@ -0,0 +1,73 @@
+{ pkgs, stdenv, fetchurl, python3Packages, python3, makeWrapper
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
{ lib, fetchFromGitHub, python3Packages, makeWrapper, sox, flac
```

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    netifaces
+    requests
+    pkgs.vorbis-tools
+    pkgs.sox
+    pkgs.flac
+    pkgs.faac
+    pkgs.lame
+    pkgs.ffmpeg
+    pkgs.nodejs
+  ] ++ stdenv.lib.optional alsaSupport [ pkgs.alsaLib pkgs.alsaUtils pkgs.ffmpeg ]
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]: ffmpeg is already unconditionally a dependency

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    ++ stdenv.lib.optional (!alsaSupport) pkgs.pulseaudio
+    ++ stdenv.lib.optional systemTray python3Packages.pyqt5
+    ++ stdenv.lib.optional ffmpegSupport pkgs.ffmpeg
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]: ffmpeg is already unconditionally a dependency

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    install -Dm 644 -t $out/lib/mkchromecast mkchromecast/*.py
+    install -Dm 644 -t $out/lib/mkchromecast/getch mkchromecast/getch/*.py
+    install -Dm 644 nodejs/html5-video-streamer.js $out/lib/nodejs/html5-video-streamer.js
+    install -Dm 644 -t $out/share/images/ images/google*.png
+    install -Dm 644 images/mkchromecast.xpm $out/share/pixmaps/mkchromecast.xpm
+    install -Dm 755 mkchromecast.py $out/bin/mkchromecast
+    wrapProgram "$out/bin/mkchromecast" --prefix PYTHONPATH : "$out/lib:$out/lib/mkchromecast/getch:$PYTHONPATH" \
+  '';
+
+  meta = with stdenv.lib; {
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
  meta = with lib; {
```

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    homepage = "https://github.com/muammar/mkchromecast";
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
    homepage = https://github.com/muammar/mkchromecast;
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/mkchromecast/default.nix (hunk 4 above)
Comment (dotlambda) [reply pkgs/applications/networking/mkchromecast/default.nix]: Actually no need to patch. We can just use `makeWrapper` to add these packages to `$PATH`.

Comment (Shou) [reply pkgs/applications/networking/mkchromecast/default.nix]: As in another `--prefix` with `--prefix PATH : "$ffmpeg.outPath/bin:$PATH"`? What's the best way to go about doing this?

**Comment by Shou** [comment]
Should've addressed the CR feedback now. Let me know if there's anything I need to change

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    faac
+    lame
+    opusTools
+    gstreamer
+    nodejs
+  ] ++ lib.optional alsaSupport [ alsaLib ffmpeg ]
+    ++ lib.optional (!alsaSupport) pulseaudio
+    ++ lib.optional systemTray py.pyqt5
+    ++ lib.optional ffmpegSupport ffmpeg
+    ++ lib.optional youtubeSupport youtube-dl;
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]: All non-Python packages don't belong in here and should (conditionally) be added to `$PATH`.

Comment (Shou) [reply pkgs/applications/networking/mkchromecast/default.nix]: I should've solved the conditional part now. Are there packages here that don't belong entirely in the PATH (and which ones)?

Comment (Shou) [reply pkgs/applications/networking/mkchromecast/default.nix]: Also cleaned up by using `lib.strings.makeSearchPathOutput` now

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    owner = "muammar";
+    repo = pname;
+    rev = version;
+    url = "https://github.com/${owner}/${pname}/archive/${version}.tar.gz";
+    sha256 = "0baxilfalndip3fx8cj03pjkdywv1v65k41srarcdrzvm8dqca9s";
+  };
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  propagatedBuildInputs = let py = python3Packages; in [
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
  propagatedBuildInputs = with python3Packages; [
```

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    py.PyChromecast
+    py.psutil
+    py.mutagen
+    py.flask
+    py.netifaces
+    py.requests
+  ] ++ lib.optional systemTray py.pyqt5 ++ packages;
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]: `packages` can be left out completely

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+  # Relies on an old version (0.7.7) of PyChromecast unavailable in Nixpkgs.
+  doCheck = false;
+
+  # Patch a relative path referring to a Javascript file, to be absolute.
+  patchPhase = ''
+    sed -i "s_\['\./nodejs_['$out/lib/nodejs_" mkchromecast/video.py
+  '';
+
+  installPhase = with lib; let
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
  postInstall = let
```

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    path = strings.makeSearchPathOutput "" "bin" (lists.remove nodejs packages);
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
    path = lib.strings.makeSearchPathOutput "" "bin" packages;
```
You can remove `nodejs` from `packages`.
Actually, I think as short as this is the `let` isn't needed.

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    install -Dm 644 man/mkchromecast.1 $out/share/man/man1/mkchromecast.1
+    install -Dm 644 mkchromecast.desktop $out/share/applications/mkchromecast.desktop
+    install -Dm 644 -t $out/lib/mkchromecast mkchromecast/*.py
+    install -Dm 644 -t $out/lib/mkchromecast/getch mkchromecast/getch/*.py
+    install -Dm 644 nodejs/html5-video-streamer.js $out/lib/nodejs/html5-video-streamer.js
+    install -Dm 644 -t $out/share/images/ images/google*.png
+    install -Dm 644 images/mkchromecast.xpm $out/share/pixmaps/mkchromecast.xpm
+    install -Dm 755 mkchromecast.py $out/bin/mkchromecast
+    wrapProgram "$out/bin/mkchromecast" \
+      --prefix PYTHONPATH : "$out/lib:$out/lib/mkchromecast/getch:$PYTHONPATH" \
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]: Why is this line needed?

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+    env
+    install -Dm 644 man/mkchromecast.1 $out/share/man/man1/mkchromecast.1
+    install -Dm 644 mkchromecast.desktop $out/share/applications/mkchromecast.desktop
+    install -Dm 644 -t $out/lib/mkchromecast mkchromecast/*.py
+    install -Dm 644 -t $out/lib/mkchromecast/getch mkchromecast/getch/*.py
+    install -Dm 644 nodejs/html5-video-streamer.js $out/lib/nodejs/html5-video-streamer.js
+    install -Dm 644 -t $out/share/images/ images/google*.png
+    install -Dm 644 images/mkchromecast.xpm $out/share/pixmaps/mkchromecast.xpm
+    install -Dm 755 mkchromecast.py $out/bin/mkchromecast
+    wrapProgram "$out/bin/mkchromecast" \
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]: You should use `makeWrapperArgs` instead.

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    py.requests
+  ] ++ lib.optional systemTray py.pyqt5 ++ packages;
+
+  # Build instructions are macOS specific in 0.3.8.1;
+  # master has Linux build instructions but awaiting new release.
+  dontBuild = true;
+
+  format = "other";
+
+  # Relies on an old version (0.7.7) of PyChromecast unavailable in Nixpkgs.
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]: If the application only works with a certain version of a dependency, you should pin the latter using `packageOverrides`.

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+
+in python3Packages.buildPythonApplication rec {
+  pname = "mkchromecast";
+  version = "0.3.8.1";
+
+  src = fetchFromGitHub rec {
+    owner = "muammar";
+    repo = pname;
+    rev = version;
+    url = "https://github.com/${owner}/${pname}/archive/${version}.tar.gz";
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
```

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+    sha256 = "0baxilfalndip3fx8cj03pjkdywv1v65k41srarcdrzvm8dqca9s";
+  };
+
+  nativeBuildInputs = [ makeWrapper ];
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
```

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+, alsaSupport ? false
+, systemTray ? true
+, ffmpegSupport ? true
+, youtubeSupport ? true
+}:
+
+# We define non-python packages here so we can reuse the binding in
+# makeSearchPathOutput below.
+let packages = [
+      vorbis-tools
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
  vorbis-tools
```

File: pkgs/top-level/all-packages.nix [hunk 21]
```diff
@@ -13703,6 +13703,13 @@ in
 
   minio = callPackage ../servers/minio { };
 
+  mkchromecast = callPackage ../applications/networking/mkchromecast {
+    alsaSupport    = config.mkchromecast.alsaSupport or false;
+    systemTray     = config.mkchromecast.systemTray or true;
+    ffmpegSupport  = config.mkchromecast.ffmpegSupport or true;
+    youtubeSupport = config.mkchromecast.youtubeSupport or true;
```
Comment (dotlambda) [inline pkgs/top-level/all-packages.nix]: I'm not very familiar with the `config` system. However, I don't think these are useful as they can be achived using a simple `override`.

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/mkchromecast/default.nix (hunk 17 above)
Comment (Shou) [reply pkgs/applications/networking/mkchromecast/default.nix]: That version isn't available in Nixpkgs at all and I assume packaging said old version is out of scope here – am I right to assume that?

File: pkgs/applications/networking/mkchromecast/default.nix (hunk 15 above)
Comment (Shou) [reply pkgs/applications/networking/mkchromecast/default.nix]: I'm not sure why entirely, in the context of Nixpkgs – is there a better way? I have just tested without the getch line and it works fine; it was there because of some issues with it that might have been specific to some interactive testing I was doing in `nix-shell` but it seems to work fine with `nix-build -A` and `nix-env -i`.

File: pkgs/applications/networking/mkchromecast/default.nix (hunk 13 above)
Comment (Shou) [reply pkgs/applications/networking/mkchromecast/default.nix]: If I use this it tries to use the makefile which doesn't have an appropriate target because it's not meant for building. Should I still use `postInstall`? How do I ideally get around the issue above if so?

**Comment by Shou** [comment]
Pushed changes following most feedback above.

File: pkgs/applications/networking/mkchromecast/default.nix (hunk 17 above)
Comment (dotlambda) [reply pkgs/applications/networking/mkchromecast/default.nix]: No need to package it separately, Just use `packageOverrides` as is e.g. done in pkgs/applications/misc/khard/default.nix.

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+    sed -i "s_\['\./nodejs_['$out/lib/nodejs_" mkchromecast/video.py
+  '';
+
+  makeWrapperArgs = [
+      ''--prefix PYTHONPATH : "$out/lib:$PYTHONPATH"''
+      ''--prefix PATH : "${lib.makeSearchPathOutput "" "bin" packages}"''
+  ];
+
+  installPhase = ''
+    env
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]: What's that meant to do?

**Review by dotlambda — CHANGES_REQUESTED** [review]
With all this custom stuff this package needs, I think we'd be better off just packaging a version of mkchromecast from their master branch.

File: pkgs/applications/networking/mkchromecast/default.nix (hunk 22 above)
Comment (Shou) [reply pkgs/applications/networking/mkchromecast/default.nix]: It's leftover from debugging, it just prints the environment. I've removed it now.

**Comment by Shou** [comment]
How do we version/name a package based on a master branch? Do we add it to `all-packages.nix` as `mkchromecastMaster` and pin it against the specific commit master is at now?

**Comment by dotlambda** [comment]
We can just keep calling the attribute `mkchromecast` and set the version to `unstable-YYYY-MM-DD`, where the date is that of the commit used.

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+  # Relies on an old version (0.7.7) of PyChromecast unavailable in Nixpkgs.
+  doCheck = false;
+
+  # Patch a relative path referring to a Javascript file, to be absolute.
+  patchPhase = ''
+    sed -i "s_\['\./nodejs_['$out/lib/nodejs_" mkchromecast/video.py
+  '';
+
+  makeWrapperArgs = [
+      ''--prefix PYTHONPATH : "$out/lib:$PYTHONPATH"''
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]:
```suggestion
    ''--prefix PYTHONPATH : "$out/lib:$PYTHONPATH"''
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/mkchromecast/default.nix [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+  installPhase = ''
+    install -Dm 644 man/mkchromecast.1 $out/share/man/man1/mkchromecast.1
+    install -Dm 644 mkchromecast.desktop $out/share/applications/mkchromecast.desktop
+    install -Dm 644 -t $out/lib/mkchromecast mkchromecast/*.py
+    install -Dm 644 -t $out/lib/mkchromecast/getch mkchromecast/getch/*.py
+    install -Dm 644 nodejs/html5-video-streamer.js $out/lib/nodejs/html5-video-streamer.js
+    install -Dm 644 -t $out/share/images/ images/google*.png
+    install -Dm 644 images/mkchromecast.xpm $out/share/pixmaps/mkchromecast.xpm
+    install -Dm 755 mkchromecast.py $out/bin/mkchromecast
+    wrapProgram "$out/bin/mkchromecast"
```
Comment (dotlambda) [inline pkgs/applications/networking/mkchromecast/default.nix]: Does this line actually do anything?

**Comment by dotlambda** [comment]
However, if using master turns out to be just as complicated, I think we can with it as it is right now.

File: pkgs/applications/networking/mkchromecast/default.nix (hunk 24 above)
Comment (Shou) [reply pkgs/applications/networking/mkchromecast/default.nix]: Huh, I swear that was necessary in the beginning because it wasn't being wrapped, but it works now without that line. I've changed it now.

**Comment by Shou** [comment]
I'm happy to go ahead with this PR, and then give packaging `master` a try sometime later this week.

**Comment by dotlambda** [comment]
Alright, but you still have to change the version of PyChromecast used.

**Comment by Shou** [comment]
@dotlambda the tests here are actually I/O bound through testing an actual Chromecast Audio device ("CCA"), I don't think we want that, right? I've just added a further comment explaining why it's disabled.

(sorry for the silly diff, relevant change is on line 52)

**Comment by bobvanderlinden** [comment]
Good to see this PR already being submitted. Looks very good :D It is actually more stable and supports the latest `PyChromecast` as well. Since it has been more than a months since the request for a new version, I'd suggest using the latest master. You can find the changes in #59280.

File: pkgs/top-level/all-packages.nix (hunk 21 above)
Comment (dotlambda) [reply pkgs/top-level/all-packages.nix]: Why did you just mark this as resolved?

File: pkgs/applications/networking/mkchromecast/default.nix (hunk 17 above)
Comment (dotlambda) [reply pkgs/applications/networking/mkchromecast/default.nix]: Since "master is actually more stable and supports the latest PyChromecast as well", we should definitely override the PyChromecast version as long as we're using version 0.3.8.1.

File: pkgs/top-level/all-packages.nix (hunk 21 above)
Comment (Shou) [reply pkgs/top-level/all-packages.nix]: My bad! I did actually remove it, but some rebasing later it was back.

File: pkgs/applications/networking/mkchromecast/default.nix (hunk 17 above)
Comment (Shou) [reply pkgs/applications/networking/mkchromecast/default.nix]: Would it benefit us when the tests aren't really useful here? Could we just do this for the one targeting `master`?

**Comment by aanderse** [comment]
So can this PR be closed now as the package was already added? Any reason to keep it open @Shou?

**Comment by Shou** [comment]
@aanderse it hasn't been merged yet, the above PR was closed. I'm just waiting for feedback here

**Comment by aanderse** [comment]
@Shou Ah sorry, misread. I was just trying to triage.

@dotlambda seeing as you have had the most review on this PR is there anything outstanding from your perspective? Do you have any feedback for @Shou?

**Comment by aanderse** [comment]
@bobvanderlinden or @dtzWill Are either of you able to provide the requested feedback here?

**Comment by bobvanderlinden** [comment]
This version works for me! Tested on NixOS with older Chromecast. Command line:

```
mkchromecast --port 8080 --video -i myvideo.mp4
```

**Comment by aanderse** [comment]
Thanks @bobvanderlinden! @Shou sound good?

**Comment by Shou** [comment]
@aanderse @bobvanderlinden have done that now

**Comment by aanderse** [comment]
Thanks @Shou! @bobvanderlinden Hopefully you'll be able to test and merge this at some point... unfortunately I don't have the hardware available currently to do so.

**Comment by bobvanderlinden** [comment]
It would be a shame of this wasn't merged.

**Comment by bobvanderlinden** [comment]
@copumpkin @zimbatm @aanderse @dotlambda

**Comment by austinbutler** [comment]
Tested this with my Chromecast Audio but ran into what appears to be this issue: https://github.com/muammar/mkchromecast/issues/308. Seems like it's possibly just an issue with `mkchromecast` itself, though. It does run and connect to the Chromecast at least.
